File: scripts/secure_workflow_payload.py
#!/usr/bin/env python3
import base64
import hashlib
import json
import os
import zlib
from pathlib import Path

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

VERSION='SCW1'
KEY_CONTEXT=b'short-cloud-workflow-payload-v1\0'


def _decode(value):
    value=str(value or '')
    return base64.urlsafe_b64decode(value+'='*((4-len(value)%4)%4))


def _event_inputs():
    path=Path(os.getenv('GITHUB_EVENT_PATH',''))
    if not path.is_file():
        return {}
    try:
        event=json.loads(path.read_text(encoding='utf-8'))
    except Exception as exc:
        raise RuntimeError('Não foi possível ler os dados protegidos desta execução.') from exc
    values=event.get('inputs') or {}
    return values if isinstance(values,dict) else {}


def _secret():
    value=(os.getenv('WORKFLOW_PAYLOAD_SECRET') or os.getenv('GROQ_API_KEY') or '').strip()
    if len(value)<16:
        raise RuntimeError('A proteção dos dados do processamento não está configurada.')
    return value


def load_secure_payload():
    inputs=_event_inputs()
    request_id=str(inputs.get('request_id') or os.getenv('INPUT_REQUEST_ID') or '').strip()
    token=str(inputs.get('secure_payload') or os.getenv('SECURE_WORKFLOW_PAYLOAD') or '').strip()
    if not request_id or not token:
        raise RuntimeError('Os dados protegidos desta execução estão ausentes.')
    parts=token.split('.')
    if len(parts)!=4 or parts[0]!=VERSION:
        raise RuntimeError('Os dados protegidos desta execução são inválidos.')
    try:
        nonce,ciphertext,tag=map(_decode,parts[1:])
        key=hashlib.sha256(KEY_CONTEXT+_secret().encode('utf-8')).digest()
        compressed=AESGCM(key).decrypt(nonce,ciphertext+tag,f'short-cloud:{request_id}'.encode('utf-8'))
        raw=zlib.decompress(compressed,-zlib.MAX_WBITS)
        data=json.loads(raw.decode('utf-8'))
    except Exception as exc:
        raise RuntimeError('Não foi possível abrir os dados protegidos desta execução.') from exc
    if not isinstance(data,dict) or str(data.get('request_id') or '')!=request_id:
        raise RuntimeError('Os dados protegidos não pertencem a esta execução.')
    return data

File: lib/secure-workflow-payload.js
const crypto=require('crypto');
const zlib=require('zlib');

const VERSION='SCW1';
const KEY_CONTEXT='short-cloud-workflow-payload-v1\0';

function workflowSecret(){
  const value=String(process.env.WORKFLOW_PAYLOAD_SECRET||process.env.GROQ_API_KEY||'').trim();
  if(value.length<16)throw new Error('A proteção dos dados do processamento não está configurada.');
  return value;
}

function deriveKey(){
  return crypto.createHash('sha256').update(KEY_CONTEXT).update(workflowSecret()).digest();
}

function encode(value){
  return Buffer.from(value).toString('base64url');
}

function sealWorkflowPayload(payload,requestId){
  const id=String(requestId||'').trim();
  if(!/^[A-Za-z0-9._-]{8,100}$/.test(id))throw new Error('Identificador de processamento inválido.');
  const serialized=Buffer.from(JSON.stringify({...payload,request_id:id}),'utf8');
  if(serialized.length>180000)throw new Error('Os dados do processamento ficaram grandes demais.');
  const compressed=zlib.deflateRawSync(serialized,{level:9});
  const nonce=crypto.randomBytes(12),cipher=crypto.createCipheriv('aes-256-gcm',deriveKey(),nonce);
  cipher.setAAD(Buffer.from(`short-cloud:${id}`,'utf8'));
  const encrypted=Buffer.concat([cipher.update(compressed),cipher.final()]),tag=cipher.getAuthTag();
  const token=[VERSION,encode(nonce),encode(encrypted),encode(tag)].join('.');
  if(token.length>60000)throw new Error('Os dados protegidos do processamento ficaram grandes demais.');
  return token;
}

module.exports={sealWorkflowPayload};
